#! /bin/sh /usr/share/dpatch/dpatch-run
## 34_fix_smbclient_check_disk_smb.dpatch by Jan Wagner <waja@cyconet.org>
## patch provided by Stephane Chazelas <stephane@artesyncp.com>
##
## DP: Fixes use of smbclient (bugs.debian.org/478942)
## DP: Upstream bug is: https://sourceforge.net/tracker/?func=detail&atid=397597&aid=1986260&group_id=29880

@DPATCH@
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' nagios-plugins-1.4.14~/plugins-scripts/check_disk_smb.pl nagios-plugins-1.4.14/plugins-scripts/check_disk_smb.pl
--- nagios-plugins-1.4.14~/plugins-scripts/check_disk_smb.pl	2010-04-11 12:41:42.000000000 +0200
+++ nagios-plugins-1.4.14/plugins-scripts/check_disk_smb.pl	2010-04-11 12:42:34.000000000 +0200
@@ -25,17 +25,13 @@
 use vars qw($opt_P $opt_V $opt_h $opt_H $opt_s $opt_W $opt_u $opt_p $opt_w $opt_c $opt_a $verbose);
 use vars qw($PROGNAME);
 use lib utils.pm ;
-use utils qw($TIMEOUT %ERRORS &print_revision &support &usage);
+use utils qw($TIMEOUT %ERRORS &print_revision &support &usage &output_and_error_of);
 
 sub print_help ();
 sub print_usage ();
 
 $PROGNAME = "check_disk_smb";
 
-$ENV{'PATH'}='';
-$ENV{'BASH_ENV'}=''; 
-$ENV{'ENV'}='';
-
 Getopt::Long::Configure('bundling');
 GetOptions
 	("v"   => \$verbose, "verbose"    => \$verbose,
@@ -58,9 +54,7 @@
 
 if ($opt_h) {print_help(); exit $ERRORS{'OK'};}
 
-my $smbclient= "$utils::PATH_TO_SMBCLIENT " ;
-my $smbclientoptions= $opt_P ? "-p $opt_P " : "";
-
+my $smbclient = $utils::PATH_TO_SMBCLIENT;
 
 # Options checking
 
@@ -72,11 +66,11 @@
 my $share = $1 if ($opt_s =~ /^([-_.A-Za-z0-9]+\$?)$/);
 ($share) || usage("Invalid share: $opt_s\n");
 
-($opt_u) || ($opt_u = shift @ARGV) || ($opt_u = "guest");
-my $user = $1 if ($opt_u =~ /^([-_.A-Za-z0-9\\]+)$/);
-($user) || usage("Invalid user: $opt_u\n");
+defined($opt_u) || ($opt_u = shift @ARGV) || ($opt_u = "guest");
+my $user = $1 if ($opt_u =~ /^([-_.A-Za-z0-9\\]*)$/);
+defined($user) || usage("Invalid user: $opt_u\n");
 
-($opt_p) || ($opt_p = shift @ARGV) || ($opt_p = "");
+defined($opt_p) || ($opt_p = shift @ARGV) || ($opt_p = "");
 my $pass = $1 if ($opt_p =~ /(.*)/);
 $pass = "-N" if ($opt_p eq "");
 
@@ -162,23 +156,19 @@
 
 # Execute an "ls" on the share using smbclient program
 # get the results into $res
-if (defined($workgroup)) {
-	if (defined($address)) {
-		print "$smbclient " . "\/\/$host\/$share" ." $pass -W $workgroup -U $user $smbclientoptions -I $address -c ls\n" if ($verbose);
-		$res = qx/$smbclient "\/\/$host\/$share" $pass -W $workgroup -U $user $smbclientoptions -I $address -c ls/;
-	} else {
-		print "$smbclient " . "\/\/$host\/$share" ." $pass -W $workgroup -U $user $smbclientoptions -c ls\n" if ($verbose);
-		$res = qx/$smbclient "\/\/$host\/$share" $pass -W $workgroup -U $user $smbclientoptions -c ls/;
-	}
-} else {
-	if (defined($address)) {
-		print "$smbclient " . "\/\/$host\/$share" ." $pass -U $user $smbclientoptions -I $address -c ls\n" if ($verbose);
-		$res = qx/$smbclient "\/\/$host\/$share" $pass -U $user $smbclientoptions -I $address -c ls/;
-	} else {
-		print "$smbclient " . "\/\/$host\/$share" ." $pass -U $user $smbclientoptions -c ls\n" if ($verbose);
-		$res = qx/$smbclient "\/\/$host\/$share" $pass -U $user $smbclientoptions -c ls/;
-	}
-}
+my @cmd = (
+  $smbclient,
+  "//$host/$share",
+  "-U", "$user%$pass",
+  defined($workgroup) ? ("-W", $workgroup) : (),
+  defined($address) ? ("-I", $address) : (),
+  defined($opt_P) ? ("-p", $opt_P) : (),
+  "-c", "ls"
+);
+
+print join(" ", @cmd) . "\n" if ($verbose);
+$res = output_and_error_of(@cmd) or exit $ERRORS{"UNKNOWN"};
+
 #Turn off alarm
 alarm(0);
 
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' nagios-plugins-1.4.14~/plugins-scripts/utils.pm.in nagios-plugins-1.4.14/plugins-scripts/utils.pm.in
--- nagios-plugins-1.4.14~/plugins-scripts/utils.pm.in	2008-11-30 22:23:18.000000000 +0100
+++ nagios-plugins-1.4.14/plugins-scripts/utils.pm.in	2010-04-11 12:41:43.000000000 +0200
@@ -7,7 +7,8 @@
 
 require Exporter;
 @ISA = qw(Exporter);
-@EXPORT_OK = qw($TIMEOUT %ERRORS &print_revision &support &usage);
+@EXPORT_OK = qw($TIMEOUT %ERRORS &print_revision &support &usage
+		&output_of &output_and_error_of);
 
 #use strict;
 #use vars($TIMEOUT %ERRORS);
@@ -64,4 +65,29 @@
 	}
 }
 
+sub output_of {
+	local *CMD;
+	local $/ = undef;
+	if (open CMD, "-|", @_) {
+		return <CMD>;
+		close CMD;
+	}
+	return undef;
+}
+
+sub output_and_error_of {
+	local *CMD;
+	local $/ = undef;
+	my $pid = open CMD, "-|";
+	if (defined($pid)) {
+		if ($pid) {
+			return <CMD>;
+		} else {
+			open STDERR, ">&STDOUT" and exec @_;
+			exit(1);
+		}
+	}
+	return undef;
+}
+
 1;
